shorebird_code_push/lib/src/shorebird_updater.dart

```diff
@@ -0,0 +1,94 @@
+import 'package:shorebird_code_push/src/shorebird_updater_io.dart'
+    if (dart.library.js_interop) './shorebird_updater_web.dart';
+import 'package:shorebird_code_push/src/updater.dart';
+
+/// Signature for a function that reports download progress.
+typedef OnDownloadProgress = void Function(int received, int total);
+
+/// {@template update_exception}
+/// Thrown when an error occurs while performing an update.
+/// {@endtemplate}
+class UpdateException implements Exception {
+  /// {@macro update_exception}
+  const UpdateException(this.message);
+
+  /// The error message.
+  final String message;
+
+  @override
+  String toString() => 'UpdateException: $message';
+}
+
+/// {@template updater_state}
+/// Information about the current state of the updater.
+/// {@endtemplate}
+sealed class UpdaterState {
+  /// {@macro updater_state}
+  const UpdaterState();
+}
+
+/// {@template updater_unavailable_state}
+/// The state of the updater when the updater is not available.
+///
+/// The most common reasons for being in this state are:
+/// 1. The app is running in debug mode (Shorebird only supports release
+///    mode).
+/// 2. The app was *NOT* built using `shorebird release` and does *NOT* contain
+///    the Shorebird engine.
+/// {@endtemplate}
+class UpdaterUnavailableState extends UpdaterState {
+  /// {@macro updater_unavailable_state}
+  const UpdaterUnavailableState();
+}
+
+/// {@template updater_available_state}
+/// Information about the current state of the updater when the
+/// updater is available (e.g. the app was built with Shorebird).
+/// {@endtemplate}
+class UpdaterAvailableState extends UpdaterState {
+  /// {@macro updater_available_state}
+  const UpdaterAvailableState({
+    required this.installedPatchNumber,
+    required this.downloadedPatchNumber,
+  });
+
+  /// The patch number of the currently installed patch.
+  /// This is the patch that the app is currently running.
+  /// Returns null if no patch is installed.
+  final int? installedPatchNumber;
```

Comment:
I think I'd probably change this to "currentPatchNumber" because we print "Patch N installed" in the updater when a patch has been downloaded and inflated, but before the patch is actually running. Using installed to mean both things might be slightly confusing.

Reply:
I was just trying to be consistent with installed vs downloaded. Do you prefer currentPatchNumber and downloadedPatchNumber? Naming is hard 🤣

Reply:
I think "current" just to avoid confusion. "runningPatchNumber" might be another option, but I personally prefer current